### DISCUSSION TITLE
Modeling unique settings as dictionary or simple list?

### DISCUSSION THREAD

**Discussion by ngbrown** [ravendb/ravendb#15164] [Q&A] [answered]
Imagine a case where I'm storing a document on how to configure a multi-channel data recorder in RavenDB. These recorders can have any of the available channels configured many different ways. There is a a collection of possible channels, with each having an ID, and another collection of possible possible settings per channel type, also with IDs.

Since each channel can only appear once, and each setting within the channel can only appear once, I'm currently modeling this as a `Dictionary<string, ChannelConfig>` in the C# class.  Is this ok or does it have any negative implications for RavenDB?

```json
{
    "TestId": "Tests/1871/header",
    "ChannelConfigs": {
        "ChannelDefinitions/5": {
            "ChannelDefinitionId": "ChannelDefinitions/5",
            "ChannelLabel": "AI5",
            "SampleRate": 1000,
            "CollectData": true,
            "Label": {
                "ChannelType": {
                    "Value": 1,
                    "DisplayName": "AI"
                },
                "Label": "Water Pressure",
                "Units": "psi"
            },
            "Configs": {
                "ConfigDefinitions/22": {
                    "Value": 20,
                    "DefinitionDescription": "HiLimit",
                    "DefinitionId": "ConfigDefinitions/22"
                },
                "ConfigDefinitions/23": {
                    "Value": 4,
                    "DefinitionDescription": "LowLimit",
                    "DefinitionId": "ConfigDefinitions/23"
                }
            }
        },
        "ChannelDefinitions/8": {
            "ChannelDefinitionId": "ChannelDefinitions/8",
            "ChannelLabel": "AI8",
            "SampleRate": 100,
            "CollectData": true,
            "Label": {
                "ChannelType": {
                    "Value": 1,
                    "DisplayName": "AI"
                },
                "Label": "Flow Sensor",
                "Units": "ma"
            },
            "Configs": {}
        },
        "ChannelDefinitions/19": {
            "ChannelDefinitionId": "ChannelDefinitions/19",
            "ChannelLabel": "TC3",
            "SampleRate": 10,
            "CollectData": true,
            "Label": {
                "ChannelType": {
                    "Value": 6,
                    "DisplayName": "TC"
                },
                "Label": "Water Temp",
                "Units": "F"
            },
            "Configs": {
                "ConfigDefinitions/48": {
                    "Value": 1,
                    "DefinitionDescription": "Channel Mode",
                    "DefinitionId": "ConfigDefinitions/48"
                }
            }
        }
    }
}
```

In C#, the `Value` is modeled as an `object` since these configurations could be numbers, boolean, and occasionally strings.

My thinking was that this ensures each channel is only added and configured once.  I suppose a corollary in another schema could be storing line items from a shopping order as a dictionary by SKU or product database ID. Since the goal would be to have only one product listed in the order, and adjust the quantity being ordered if the user adds more to the shopping cart.

I initially just stored by the number part of the id (`"5"` instead of `"ChannelDefinitions/5`) but RavenDB Studio warned of issues (with order changing when editing?) when storing numbers as the keys.

Any other thoughts?

**Comment by georgiosd** [comment]
This should be fine!

The order warning is just that - order may change when you save from
Studio, but doesnt sound important in how you use them.



On Mon, 24 Oct 2022 at 21:40, Nathan Brown ***@***.***> wrote:

> Imagine a case where I'm storing a document on how to configure a
> multi-channel data recorder in RavenDB. These recorders can have any of the
> available channels configured many different ways. There is a a collection
> of possible channels, with each having an ID, and another collection of
> possible possible settings per channel type, also with IDs.
>
> Since each channel can only appear once, and each setting within the
> channel can only appear once, I'm currently modeling this as a Dictionary<string,
> ChannelConfig> in the C# class. Is this ok or does it have any negative
> implications for RavenDB?
>
> {
>     "TestId": "Tests/1871/header",
>     "ChannelConfigs": {
>         "ChannelDefinitions/5": {
>             "ChannelDefinitionId": "ChannelDefinitions/5",
>             "ChannelLabel": "AI5",
>             "SampleRate": 1000,
>             "CollectData": true,
>             "Label": {
>                 "ChannelType": {
>                     "Value": 1,
>                     "DisplayName": "AI"
>                 },
>                 "Label": "Water Pressure",
>                 "Units": "psi"
>             },
>             "Configs": {
>                 "ConfigDefinitions/22": {
>                     "Value": 20,
>                     "DefinitionDescription": "HiLimit",
>                     "DefinitionId": "ConfigDefinitions/22"
>                 },
>                 "ConfigDefinitions/23": {
>                     "Value": 4,
>                     "DefinitionDescription": "LowLimit",
>                     "DefinitionId": "ConfigDefinitions/23"
>                 }
>             }
>         },
>         "ChannelDefinitions/8": {
>             "ChannelDefinitionId": "ChannelDefinitions/8",
>             "ChannelLabel": "AI8",
>             "SampleRate": 100,
>             "CollectData": true,
>             "Label": {
>                 "ChannelType": {
>                     "Value": 1,
>                     "DisplayName": "AI"
>                 },
>                 "Label": "Flow Sensor",
>                 "Units": "ma"
>             },
>             "Configs": {}
>         },
>         "ChannelDefinitions/19": {
>             "ChannelDefinitionId": "ChannelDefinitions/19",
>             "ChannelLabel": "TC3",
>             "SampleRate": 10,
>             "CollectData": true,
>             "Label": {
>                 "ChannelType": {
>                     "Value": 6,
>                     "DisplayName": "TC"
>                 },
>                 "Label": "Water Temp",
>                 "Units": "F"
>             },
>             "Configs": {
>                 "ConfigDefinitions/48": {
>                     "Value": 1,
>                     "DefinitionDescription": "Channel Mode",
>                     "DefinitionId": "ConfigDefinitions/48"
>                 }
>             }
>         }
>     }
> }
>
> In C#, the Value is modeled as an object since these configurations could
> be numbers, boolean, and occasionally strings.
>
> My thinking was that this ensures each channel is only added and
> configured once. I suppose a corollary in another schema could be storing
> line items from a shopping order as a dictionary by SKU or product database
> ID. Since the goal would be to have only one product listed in the order,
> and adjust the quantity being ordered if the user adds more to the shopping
> cart.
>
> I initially just stored by the number part of the id ("5" instead of
> "ChannelDefinitions/5) but RavenDB Studio warned of issues (with order
> changing when editing?) when storing numbers as the keys.
>
> Any other thoughts?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/ravendb/ravendb/discussions/15164>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAMYG5RWSOZAEZWJXCOSXDDWE3JZNANCNFSM6AAAAAARNH3D5E>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>
-- 
Georgios Diamantopoulos
Founder
Zero to MVP
https://zerotomvp.dev

**Comment by ayende** [comment, marked as answer]
From RavenDB's perspective, there is no real difference between the two options.
If you are doing a lot of lookups by channel id, may save you scanning the list, but that is a client side issue, not related to RavenDB.
